### PR TITLE
fix unnecessary-comprehension provides incorrect suggestion

### LINKED ISSUE #9109: unnecessary-comprehension provides incorrect suggestion
### Bug description

`unnecessary-comprehension` provides an incorrect suggestion when the collection is modified after the new list is created. The result of a comprehension is a new object and hence the suggestion should also always result in a new object to avoid code regressions.

I saved the following code in a file called `test.py` to produce the output below.

```python
#pylint: disable=missing-module-docstring
a = [1, 2, 3]
b = [x for x in a]
b[0] = 0
print(a) # [1, 2, 3]
```

After changing `b = [x for x in a]` to `b = a` based on the suggestion, the script now prints `[0, 2, 3]`. The correct suggestion should be `use list(a)` to preserve the original behavior.

### Configuration

_No response_

### Command used

```shell
pylint test.py
```

### Pylint output

```python
$ pylint test.py
************* Module test
test.py:3:4: R1721: Unnecessary use of a comprehension, use a instead. (unnecessary-comprehension)
```

### Expected behavior

```shell
$ pylint test.py
************* Module test
test.py:3:4: R1721: Unnecessary use of a comprehension, use list(a) instead. (unnecessary-comprehension)
```

### Pylint version

```shell
pylint 2.17.6
astroid 2.15.7
Python 3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)]
```

### OS / Environment

Windows 10

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by C0DE-SLAYER**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Closes #9109

### CODE REVIEW

**Comment by jacobtylerwalls** [comment]
Thanks for the patch!

Based on the test results, you'll notice you'll need to adjust the logic of #4500 to make it a little dumber (and therefore more like the original intent of #2999).

**Comment by C0DE-SLAYER** [comment]
@jacobtylerwalls i am a beginner and having a hard time solving can you help me what i have to do to solve the issue. it would be a great help! thanks

**Comment by Pierre-Sassoulas** [comment]
I'm on mobile si I can't link the specific code but  the message is raised in the add_message function, the message is dynamic and the argument given to add_message used in %s later is what need to be modified.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for your contribution, it looks promising, do you mind adding the functional tests from #8577 
```python
a = [[1, 2], [3, 4]]
b = [(x, y) for x, y in a]  # flagged as unnecessary-comprehension
```
And #9109:
```python
#pylint: disable=missing-module-docstring
a = [1, 2, 3]
b = [x for x in a]
b[0] = 0
print(a) # [1, 2, 3]
```
in tests/functional/u/unnecessary/unnecessary_comprehension.py, please ?

**Comment by Pierre-Sassoulas** [comment]
Also a bug / or false positive changelog fragment would be nice for this one.

**Comment by C0DE-SLAYER** [comment]
> Thank you for your contribution, it looks promising, do you mind adding the functional tests from #8577
> 
> ```python
> a = [[1, 2], [3, 4]]
> b = [(x, y) for x, y in a]  # flagged as unnecessary-comprehension
> ```
> 
> And #9109:
> 
> ```python
> #pylint: disable=missing-module-docstring
> a = [1, 2, 3]
> b = [x for x in a]
> b[0] = 0
> print(a) # [1, 2, 3]
> ```
> 
> in tests/functional/u/unnecessary/unnecessary_comprehension.py, please ?

in the unnecessary_comprehension.py on line 5 it can be use for 9109 and 
for 8577  ' [(x, y) for x, y in iterable]  # [unnecessary-comprehension] ' is this good

**Comment by jacobtylerwalls** [comment]
> where to do that in the doc/whatsnew/fragment or any where else and what it would be helpfull thanks!

Check out the [contributor docs](https://pylint.readthedocs.io/en/latest/development_guide/contributor_guide/contribute.html#creating-a-pull-request) for some info on documenting your pull request.

**Review by jacobtylerwalls — COMMENTED** [review]
Thanks for the news entry! Spotted one typo.

File: doc/whatsnew/fragments/9172.false_positive [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Fixed `unnecessary-comprehension provides incorrect suggestion`
```
[the remark this answers is not in the document]
Comment (C0DE-SLAYER) [reply doc/whatsnew/fragments/9172.false_positive:1]: Fixed incorrect suggestion for shallow copy in unnecessary-comprehension
does this look good

**Comment by jacobtylerwalls** [comment]
Thanks for the contribution! :gift:

**Review by jacobtylerwalls — APPROVED** [review]

**Pull request merged**